src/bulls_and_cows_lib/game_solver.cpp

```diff
@@ -1,7 +1,89 @@
 
 #include "game_solver.hpp"
+#include <algorithm>
 
 namespace bulls_and_cows {
-    // TODO: define the body of the functions declared in game_solver.cpp
 
-} // namespace bulls_and_cows
+	void code_gen(const GameOptions& game_options, PossibleSolutions& pls, std::string used_alphabet,
+		Code& code)
+	{
+		for (char temp : used_alphabet)
+		{
+			//code creation, step by step
+			if (code.value.length() < game_options.number_of_characters_per_code)
+			{
+				code.value.push_back(temp);
+
+				code_gen(game_options, pls, used_alphabet, code);
+				code.value.pop_back(); 
+			}
+			//push in vector
+			else if (code.value.length() >= game_options.number_of_characters_per_code)
+			{
+				pls.codes.push_back(code);
+				break;
+			}
+		}
+	}
+
+	void get_code_without_duplicates(PossibleSolutions& pls)
+	{
+		pls.codes.erase(std::remove_if(pls.codes.begin(), pls.codes.end(),
+			[](Code const& code) -> bool {
+				std::string used_chars;
+				for (char c : code.value)
+					if (used_chars.find(c) != std::string::npos)
+					{
+						return true;
+					}
+					else
+					{
+						used_chars.push_back(c);
+					}
+				return false;
+			}),
+			pls.codes.end());
+	}
+
+	PossibleSolutions generate_all_possible_codes(const GameOptions& game_options)
+	{
+		PossibleSolutions avc{};
+		std::string used_alphabeat;
+		for (char c = game_options.minimum_allowed_character; c <= game_options.maximum_allowed_character; c++)
+		{
+			used_alphabeat.push_back(c);
+		}
+		Code codes;
+		code_gen(game_options, avc, used_alphabeat, codes);
+
+		if (game_options.isDuplicate == false) {
+			get_code_without_duplicates(avc);
+		}
+
+		return avc;
+	}
+
+
+	Code pick_random_attempt(const PossibleSolutions& possible_solutions)
+	{
+		if (possible_solutions.codes.size() != 0) {
+			int pos = (generate_random_integer(0, static_cast<int> (possible_solutions.codes.size() - 1)));
+			return possible_solutions.codes[pos];
+		}
```

Comment:
Que ferai la ligne suivante si possible_solutions.codes.size() == 0 ?

Reply:
Ce serait une défaite dans ce cas non?

Reply:
Je ne pose pas la question d'un point de vue fonctionnel, mais d'un point de vue technique:
Si tu as un vector v dont la taille est nulle et que tu accèdes à v[0], que se passe-t-il ?

Reply:
J'imagine qu'une exception est levée par la librairie standard, comme un out of bounds ou quelque chose comme ça

Reply:
Non, c'est le problème : le comportement de ton programme si tu utilises v[0] pour un vector vide est "UNDEFINED BEHAVIOR". Tout peut arriver : un crash, un programme qui semble fonctionner mais qui utilise des données corrompues, ... C'est le prix à payer pour que l'appel de v[0] soit le + rapide possible : ce n'est pas safe.
Si tu veux une version qui lève une exception lorsque tu accèdes à l'élément 0, il faut utiliser la fonction membre at(index).

Reply:
Ah oui je vois, très bien, je change tout ça. Merci.